src/data/pnat/basic.lean

```diff
@@ -3,39 +3,35 @@ Copyright (c) 2017 Microsoft Corporation. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Mario Carneiro, Neil Strickland
 -/
+import data.pnat.defs
 import data.nat.order
 import algebra.order.positive.ring
 
 /-!
 # The positive natural numbers
 
-This file defines the type `ℕ+` or `pnat`, the subtype of natural numbers that are positive.
+This file develops the type `ℕ+` or `pnat`, the subtype of natural numbers that are positive.
+It is defined in `data.pnat.defs`, but most of the development is deferred to here so
+that file can have very few imports.
 -/
 
-/-- `ℕ+` is the type of positive natural numbers. It is defined as a subtype,
-  and the VM representation of `ℕ+` is the same as `ℕ` because the proof
-  is not stored. -/
-@[derive [decidable_eq, add_left_cancel_semigroup, add_right_cancel_semigroup, add_comm_semigroup,
-  linear_ordered_cancel_comm_monoid, linear_order, has_add, has_mul, has_one, distrib]]
-def pnat := {n : ℕ // 0 < n}
-notation `ℕ+` := pnat
-
-instance coe_pnat_nat : has_coe ℕ+ ℕ := ⟨subtype.val⟩
-instance : has_repr ℕ+ := ⟨λ n, repr n.1⟩
+-- We use `dsimp_result` here to avoid extraneous appearances of `id` in the term.
+instance : add_left_cancel_semigroup ℕ+ := by dsimp_result { dsimp [pnat], apply_instance, }
+instance : add_right_cancel_semigroup ℕ+ := by dsimp_result { dsimp [pnat], apply_instance, }
+instance : add_comm_semigroup ℕ+ := by dsimp_result { dsimp [pnat], apply_instance, }
+instance : linear_ordered_cancel_comm_monoid ℕ+ := by dsimp_result { dsimp [pnat], apply_instance, }
+instance : has_add ℕ+ := by dsimp_result { dsimp [pnat], apply_instance, }
+instance : has_mul ℕ+ := by dsimp_result { dsimp [pnat], apply_instance, }
+instance : distrib ℕ+ := by dsimp_result { dsimp [pnat], apply_instance, }
```

Comment:
Why not `attribute [derive the_same_list_as_before] pnat`?

Reply:
Ahha! Somehow I forgot that one could use `derive` after the fact, I'm an idiot. Thanks. :-)